waku/v2/peermanager/peer_manager.go

```diff
@@ -241,6 +243,98 @@ func (pm *PeerManager) Start(ctx context.Context) {
 		go pm.peerEventLoop(ctx)
 	}
 	go pm.connectivityLoop(ctx)
+	go pm.peerStoreLoop(ctx)
+}
+
+func (pm *PeerManager) peerStoreLoop(ctx context.Context) {
+	t := time.NewTicker(prunePeerStoreInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			pm.prunePeerStore()
+		}
+	}
+}
+
+func (pm *PeerManager) prunePeerStore() {
+	peers := pm.host.Peerstore().Peers()
+	numPeers := len(peers)
+	if numPeers < pm.maxPeers {
+		return
+	}
+	peerCntBeforePruning := numPeers
+	pm.logger.Debug("peerstore capacity exceeded, hence pruning", zap.Int("capacity", pm.maxPeers), zap.Int("numPeers", peerCntBeforePruning))
+
+	for _, peerID := range peers {
+		connFailues := pm.host.Peerstore().(wps.WakuPeerstore).ConnFailures(peerID)
+		if connFailues > maxFailedAttempts {
+			// safety check so that we don't end up disconnecting connected peers.
+			if pm.host.Network().Connectedness(peerID) == network.Connected {
+				pm.host.Peerstore().(wps.WakuPeerstore).ResetConnFailures(peerID)
+				continue
+			}
+			pm.host.Peerstore().RemovePeer(peerID)
+			numPeers--
+		}
+		if numPeers < pm.maxPeers {
+			pm.logger.Debug("finished pruning peer store", zap.Int("capacity", pm.maxPeers), zap.Int("beforeNumPeers", peerCntBeforePruning), zap.Int("afterNumPeers", numPeers))
+			return
+		}
+	}
+
+	notConnectedPeers := pm.getPeersBasedOnconnectionStatus("", network.NotConnected)
+	peersByTopic := make(map[string]peer.IDSlice)
+
+	if pm.RelayEnabled {
+		//prune not connected peers without shard
+		for _, peerID := range notConnectedPeers {
+			topics, err := pm.host.Peerstore().(wps.WakuPeerstore).PubSubTopics(peerID)
+			//Prune peers without pubsubtopics.
+			if err != nil || len(topics) == 0 {
```

Comment:
Should we log the error? Looking at the implementation of `PubsubTopics` in go-waku and `Get` in go-libp2p, seems to me that getting an error should be rare, so perhaps we might want to log it?

Reply:
yea, error should only happen if fetch failed which is very rare.
nonetheless better to log it since it is anyways rare.